### ISSUE TITLE
Bug: Java installation failing

### ISSUE THREAD

**Issue by aksh1618** [sdkman/sdkman-cli#987]
**Bug report**
Installation of java is failing with error:
```
mv: target '/home/aakarshituppal/.sdkman/candidates/java/8.0.302-open' is not a directory
```
Tried for versions 8.0.302-open, 11.0.12-open, 8.302.08.1-amzn

**To reproduce**
```
❯ sdk install java 8.0.302-open

Found a previously downloaded java 8.0.302-open archive. Not downloading it again...

Installing: java 8.0.302-open
mv: target '/home/aakarshituppal/.sdkman/candidates/java/8.0.302-open' is not a directory
Done installing!
```

**System info**
- OS: Ubuntu 20.04
- Shell: 
  - zsh 5.8 (x86_64-ubuntu-linux-gnu)
  - GNU bash, version 5.0.17(1)-release
- The output of `sdk version`: SDKMAN 5.12.4
-->

**Comment by marc0der** [comment]
Hi @aksh1618,

I've never seen this happening before, and I can't replicate it either. You might need to give more context about how this came about.

These are the steps I took to see if I could replicate it:

```
marco@anubis % sdk install java 8.0.302-open                                                                                                                                                                       
Downloading: java 8.0.302-open
In progress...
Repackaging Java 8.0.302-open...
Done repackaging...
Installing: java 8.0.302-open
Done installing!
Do you want java 8.0.302-open to be set as default? (Y/n): n

marco@anubis % sdk u java 8.0.302-open                                                                                                                                                                             
Using java version 8.0.302-open in this shell.

marco@anubis [1] % java -version                                                                                                                                                                                   openjdk version "1.8.0_302"
OpenJDK Runtime Environment (build 1.8.0_302-b08)
OpenJDK 64-Bit Server VM (build 25.302-b08, mixed mode)

marco@anubis % sdk uninstall java 8.0.302-open                                                                                                                                                                     
Uninstalling java 8.0.302-open...

marco@anubis % sdk install java 8.0.302-open                                                                                                                                                                       
Found a previously downloaded java 8.0.302-open archive. Not downloading it again...
Installing: java 8.0.302-open
Done installing!
```

**Comment by aksh1618** [comment]
@marc0der From what I understand, if I run, say, 
```sh
sdk install java 8.302.08.1-amzn
```
this is what's happening in `__sdkman_install_candidate_version`

1. Target directory is created
    ```sh
    mkdir -p "/home/aakarshituppal/.sdkman/candidates/java"
    ```
2. Archive is unzipped into fresh tmp directory
    ```sh
    rm -rf "/home/aakarshituppal/.sdkman/tmp/out"
    unzip -oq "/home/aakarshituppal/.sdkman/archives/java-8.302.08.1-amzn.zip" -d "/home/aakarshituppal/.sdkman/tmp/out"
    ```
3. Now at this stage the target contents are in the `out` directory. Now we have 2 options, either to create the version directory in the candidate directory, and move all the contents:
    ```sh
    mkdir "/home/aakarshituppal/.sdkman/candidates/java/8.302.08.1-amzn" # This step is missing
    mv -f "/home/aakarshituppal/.sdkman"/tmp/out/* "/home/aakarshituppal/.sdkman/candidates/java/8.302.08.1-amzn"
    ```
    or, to move the directory itself there:
    ```sh
    # First argument of mv command changed
    mv -f "/home/aakarshituppal/.sdkman"/tmp/out "/home/aakarshituppal/.sdkman/candidates/java/8.302.08.1-amzn"
    ```

**Comment by marc0der** [comment]
Hi @aksh1618,

Every archive contains an enclosing folder, such as `jdk-11.0.11`. It is this folder that we move under the candidate directory. This is standard across all SDK archives. It also guarantees that we don't leak files that might already be under `tmp/out`.

Let's first determine why you are seeing this error, as nobody else has had this happening up to now.

Firstly, could I ask you to flush your archives folder and try to replicate this again?

**Comment by aksh1618** [comment]
@marc0der Tried flushing archives, still same error. Also, here's the contents of the `out` directory after the unzip step:
```
❯ ll /home/aakarshituppal/.sdkman/tmp/out/
total 44K
drwxr-xr-x  7 aakarshituppal aakarshituppal 4.0K Jul 16 01:00 amazon-corretto-8.302.08.1-linux-x64
drwxrwxr-x  2 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 bin
drwxrwxr-x  4 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 conf
drwxrwxr-x  4 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 demo
drwxrwxr-x  3 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 include
drwxrwxr-x  2 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 jmods
drwxrwxr-x 73 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 legal
drwxrwxr-x  6 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 lib
drwxrwxr-x  4 aakarshituppal aakarshituppal 4.0K Jul 14 22:38 man
drwxrwxr-x  9 aakarshituppal aakarshituppal 4.0K Jul 16 14:51 openjdk-8u302-b08
-rw-rw-r--  1 aakarshituppal aakarshituppal 1.2K Jul 14 22:38 release
```

**Comment by marc0der** [comment]
The `openjdk-8u302-b08` directory will be zipped up and placed in the archives folder. This should be the _only thing_ under `tmp/out`. You seem to have a lot of stuff in your `tmp/out` folder which doesn't belong there, I'd suggest doing a complete flush before doing anything else.

Also, have you made any local changes/hacks to the sdkman scripts or is this a clean installation? You will see that the step before the unzip actually does a forced removal of `tmp/out` to ensure that is clean.

**Comment by aksh1618** [comment]
@marc0der I uninstalled and reinstalled sdkman:
```sh
cd
tar zcvf ~/sdkman-backup_$(date +%F-%kh%M).tar.gz -C ~/ .sdkman
rm -rf ~/.sdkman
curl -s "https://get.sdkman.io" | bash
```

And then tried installing it again, but facing same error:
```
❯ sdk install java 8.0.302-open
zsh: correct 'java' to '.java' [nyae]? n
==== BROADCAST =================================================================
* 2021-09-28: jreleaser 0.7.0 available on SDKMAN! https://github.com/jreleaser/jreleaser/releases/tag/v0.7.0
* 2021-09-24: micronaut 3.0.2 available on SDKMAN!
* 2021-09-23: springboot 2.5.5 available on SDKMAN!
================================================================================

Downloading: java 8.0.302-open

In progress...

############################################################################################################################################### 100.0%############################################################################################################################################### 100.0%

Repackaging Java 8.0.302-open...

Done repackaging...

Installing: java 8.0.302-open
mv: target '/home/aakarshituppal/.sdkman/candidates/java/8.0.302-open' is not a directory
Done installing!


Setting java 8.0.302-open as default.
```

These are the contents of the archive generated:
```
❯ zipinfo -1 /home/aakarshituppal/.sdkman/archives/java-8.0.302-open.zip | awk -F/ '{print $1}' | uniq
ASSEMBLY_EXCEPTION
man
include
jre
THIRD_PARTY_README
release
bin
sample
src.zip
demo
lib
LICENSE
```

**Comment by marc0der** [comment]
Okay, this is definitely something to do with your local shell config.
Trying it with a fresh untainted environment:

```
$ docker run -it --rm ubuntu:latest bash
# apt update && apt install -y zip unzip curl zsh
# zsh
# curl -s https://get.sdkman.io | bash
# source "/root/.sdkman/bin/sdkman-init.sh"
# sdk install java 8.0.302-open
Downloading: java 8.0.302-open
In progress...
################################################################################################## 100.0%
Repackaging Java 8.0.302-open...
Done repackaging...
Installing: java 8.0.302-open
Done installing!
Setting java 8.0.302-open as default.
# java -version
openjdk version "1.8.0_302"
OpenJDK Runtime Environment (build 1.8.0_302-b08)
OpenJDK 64-Bit Server VM (build 25.302-b08, mixed mode)
# unzip -l /root/.sdkman/archives/java-8.0.302-open.zip
Archive:  /root/.sdkman/archives/java-8.0.302-open.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2021-07-16 09:21   openjdk-8u302-b08/
    19274  2021-07-16 09:21   openjdk-8u302-b08/LICENSE
        0  2021-07-16 09:21   openjdk-8u302-b08/include/
    74698  2021-07-16 09:21   openjdk-8u302-b08/include/jni.h
     7404  2021-07-16 09:21   openjdk-8u302-b08/include/jdwpTransport.h
...
```
As you can see, no problem whatsoever. My guess is that you have some aliases defined in your shell that are introducing this odd behaviour.

**Comment by aksh1618** [comment]
@marc0der Yes, that seems to be the only explanation, I'll try to debug further and find the culprit over the weekend.

**Comment by marc0der** [comment]
Let us know how it goes. I'm curious to hear what it was.

**Comment by aksh1618** [comment]
@marc0der I've found the culprit, it was this snippet in my zsh config:
```zsh
function chpwd() {
    files=$(ls -A | wc -l)
    if [[ $files = "1" ]]; then
        zmodload zsh/parameter
        if [[ ".." != $history[$HISTCMD] ]]; then
            f=$(ls -A)
            if [[ -d "$f" ]]; then
                cd "$f"
            fi
        fi
    fi
}
```
The purpose of this function is to skip single subdirectory nested directories while using `cd`. This was making the command
```sh
cd "$work_dir"
```
in the post-installation hook (`/home/aakarshituppal/.sdkman/tmp/hook_post_java_17-open.sh`) cd to `/home/aakarshituppal/.sdkman/tmp/out/jdk-17` instead of `/home/aakarshituppal/.sdkman/tmp/out`, leading to the incorrect packaging with the following `zip` command. Verified working fine on commenting this function. Thanks for the prompt support and making this amazing tool!
